### PR TITLE
DrivAerML: 600 batches/epoch + gradient clipping (max_norm=1.0)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

Compound the 600-batch improvement (11.97% val) with gradient clipping (max_norm=1.0) to stabilize DrivAerML training. Askeladd's multi-seed experiment revealed extreme training instability on DrivAerML (seed=1 → 43.93% vs default → 12.70%). Gradient clipping should stabilize training and reduce variance, potentially pushing below 11.97%. Gradient clipping is a proven technique for taming rugged optimization landscapes — it prevents large gradient steps from destabilizing well-found minima.

## Instructions

Run the following command exactly:

```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --grad-clip 1.0 \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 600 --max-eval-batches 200 \
  --agent shoya --wandb-name "shoya/drivaerml-600batch-gradclip" \
  --wandb-group "drivaerml-600batch-compounds"
```

Key change vs baseline: added `--grad-clip 1.0` to the 600-batches/epoch configuration.

Report the best `val_primary/surface_rel_l2_pct` and the epoch it was achieved. Also report total epochs completed within the timeout.

## Baseline

Current best DrivAerML metrics (PR #2645, taki — 600 batches/epoch):
- **val_primary/surface_rel_l2_pct:** 11.97% ← target to beat
- **test_primary/surface_rel_l2_pct:** 13.03%
- **W&B run:** dar47nwl (34 epochs, 30-min timeout)
- **Reproduce baseline:** `cd target/icml2026 && python train.py --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine_t_max 30 --no-use-ema --enable-fourier --model-layers 4 --model-hidden-dim 256 --model-heads 4 --epochs 999 --batch-size 1 --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 --max-train-batches 600 --max-eval-batches 200`

**CRITICAL flags (required to avoid OOM):**
- `--batch-size 1`
- `--drivaerml-train-surface-points 50000`
- `--drivaerml-eval-surface-points 50000`
- `--max-train-batches 600`
- `--max-eval-batches 200`
- `--model-heads 4` (required for 4L/256d)

External target: <3.71% (AB-UPT) — 3.2x gap remaining.

### CODE REVIEW

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2676 because ADVISOR: Closing — DrivAerML baseline shifted to 5.73% (4L/384d, PR #2602). The 4L/256d+600batch config cannot compete. Redirecting to 4L/384d experiments..

**Comment by tcapelle** [comment]
STUDENT shoya:

## Results

### DrivAerML: 600 batches/epoch + gradient clipping (max_norm=1.0)

| Metric | Baseline (PR #2645) | This Run | Delta |
|--------|-------------------|----------|-------|
| **val_primary/surface_rel_l2_pct** | **11.97%** | **12.71%** | +0.74% ❌ |
| test_primary/surface_rel_l2_pct | 13.03% | 13.33% | +0.30% ❌ |
| Epochs completed | 34 | 34 | — |

### Details

- **Best val_primary/surface_rel_l2_pct:** 12.71% at epoch 33
- **test_primary/surface_rel_l2_pct:** 13.33%
- **W&B run:** [`tqru35ix`](https://wandb.ai/wandb-applied-ai-team/senpai-v1/runs/tqru35ix)
- **Total epochs:** 34 (30-min timeout)
- **Peak memory:** ~50K surface points, batch_size=1 on single RTX PRO 6000 Blackwell (97GB VRAM)

**Command used:**
```bash
cd target/icml2026 && CUDA_VISIBLE_DEVICES=0 python train.py \
  --dataset drivaerml --optimizer adamw --lr 5e-4 --cosine-t-max 30 \
  --no-use-ema --enable-fourier --grad-clip 1.0 \
  --model-layers 4 --model-hidden-dim 256 --model-heads 4 \
  --epochs 999 --batch-size 1 \
  --drivaerml-train-surface-points 50000 --drivaerml-eval-surface-points 50000 \
  --max-train-batches 600 --max-eval-batches 200 \
  --agent shoya --wandb-name "shoya/drivaerml-600batch-gradclip" \
  --wandb-group "drivaerml-600batch-compounds"
```

### What happened

Gradient clipping (max_norm=1.0) **did not beat the baseline**. The best val metric was 12.71% vs 11.97% baseline — 0.74% worse.

Key observations:
- **Gradient norms were consistently above the clip threshold** for the first ~27 epochs (starting at ~5.0, declining to ~1.0 by epoch 27). This means clipping was actively constraining gradients for most of training.
- **Convergence was slower** — the model was still improving at epoch 34 when the timeout hit, whereas the baseline (without clipping) reached 11.97% in the same number of epochs.
- The clipping appears to have **slowed convergence without providing stability benefits** in this case. The baseline training (without clipping) was already stable enough to reach a better final metric.
- Grad norm decreased smoothly from ~5.0 → ~0.79 over 34 epochs, suggesting training was already reasonably stable without clipping.

### Suggested follow-ups

- **Try a higher clip threshold** (e.g., max_norm=5.0 or 10.0) that only clips extreme outlier gradients rather than constraining normal training
- **Try gradient clipping with a higher LR** — clipping could be more beneficial when paired with aggressive learning rates where instability is more likely
- The multi-seed variance issue (43.93% on seed=1) may be better addressed by other regularization techniques (dropout, weight decay) rather than gradient clipping

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2676 because ADVISOR: Closing — 4L/256d experiments are obsolete after the 4L/384d breakthrough (5.73%). This PR's 12.82% at 4L/256d cannot compete. shoya, please pick up your next assignment..
